iwyu.cc

```diff
@@ -4099,6 +4132,31 @@ class InstantiatedTemplateVisitor
     return TraverseDecl(const_cast<FunctionDecl*>(fn_decl));
   }
 
+  bool TraverseExpandedTemplateVariableHelper(VarDecl* decl) {
+    if (!decl || ContainsKey(traversed_decls_, decl))
+      return true;  // avoid recursion and repetition
+    // TODO(bolshakov): is this really needed? BaseAstVisitor::Traverse* methods
+    // already have protection from recursion.
+    traversed_decls_.insert(decl);
+
+    AstFlattenerVisitor nodeset_getter(compiler());
+    // This gets to the decl for the (uninstantiated) template-as-written:
```

Comment:
This is significantly simpler than the recursive push/pop thing going on in TraverseExpandedTemplateFunctionHelper: https://github.com/include-what-you-use/include-what-you-use/pull/1786/files#diff-7ca107c8b836c1765e8b735b74f06833bec48f308a76017886cee1d5736002d9L4076.

I have a very shallow understanding of variable templates, but are they just less nested than function templates by definition? Or why is the `ValueSaver` not necessary here?

Reply:
I've found in the history that `ValueSaver` was added there by me for a very specific task: so as not to block the second analysis of a function from `HandleFunctionCall` after the first one has been made from `VisitDeclRefExpr`. It is important for reporting the function return type from `HandleFunctionCall`. This is irrelevant to variable templates. But I can add `ValueSaver` here as well to be consistent and maybe invent some specific test case, despite I doubt that it would be useful in real life...

Reply:
In fact, I have an idea that `AstFlattenerVisitor` should be replaced some day with some more advanced approach to block reporting of entities used at the template definition side... It would be needed to report types not only from template arguments.

Reply:
Thanks for explaining. I think it's good as-is without `ValueSaver` if it's just there for a specific function-call quirk.